### PR TITLE
Make course search faster using FlexSearch

### PR DESCRIPTION
**Pull request by SamZhang02**
Changed the main page's search mechanism from requesting the backend requests to only frontend FlexSearch.

### CODE REVIEW

File: package-lock.json
Comment (terror) [inline package-lock.json:1]: bruh

Comment (SamZhang02) [reply package-lock.json:1]: 😭 gotta stop `git add .`

File: client/src/lib/utils.ts [hunk 1]
```diff
@@ -88,3 +88,18 @@ export const spliceCourseCode = (courseCode: string, delimiter: string) =>
...
+
+  arr.forEach((item) => {
+    const key = by(item);
+    if (!unique.has(key)) {
+      unique.set(key, item);
+    }
+  });
+
+  return Array.from(unique.values());
+};
```
Comment (JeffDotPng) [inline client/src/lib/utils.ts:105]: just use lodash instead it already has `_.uniq` and `_.uniqBy`

File: client/src/pages/Home.tsx [hunk 2]
```diff
@@ -1,16 +1,46 @@
...
+
+courses.forEach((course, i) =>
+  coursesIndex.add(
+    i,
+    `${course._id} ${course.subject} ${course.title} ${course.code}`
+  )
+);
+instructors.forEach((instructor, i) =>
+  instructorsIndex.add(i, instructor.name)
+);
```
Comment (JeffDotPng) [inline client/src/pages/Home.tsx:42]: don't you need to update the course search bar in the header as well? just put this in some file under `/lib`

**Comment by SamZhang02** [comment]
lookin nice :100:

**Review by JeffDotPng — APPROVED** [review]

**Pull request merged**